src/transformation/utils/diagnostics.ts

```diff
@@ -3,126 +3,146 @@ import { LuaTarget } from "../../CompilerOptions";
 );
 
-export const luaTableCannotBeAccessedDynamically = createDiagnosticFactory("@luaTable cannot be accessed dynamically.");
+export const luaTableCannotBeAccessedDynamically = createErrorDiagnosticFactory(
+    "@luaTable cannot be accessed dynamically."
+);
 
-export const luaTableForbiddenUsage = createDiagnosticFactory(
+export const luaTableForbiddenUsage = createErrorDiagnosticFactory(
     (description: string) => `Invalid @luaTable usage: ${description}.`
 );
 
-export const luaIteratorForbiddenUsage = createDiagnosticFactory(
+export const luaIteratorForbiddenUsage = createErrorDiagnosticFactory(
     "Unsupported use of lua iterator with '@tupleReturn' annotation in for...of statement. " +
         "You must use a destructuring statement to catch results from a lua iterator with " +
         "the '@tupleReturn' annotation."
 );
 
-export const unsupportedAccessorInObjectLiteral = createDiagnosticFactory(
+export const unsupportedAccessorInObjectLiteral = createErrorDiagnosticFactory(
     "Accessors in object literal are not supported."
 );
 
-export const unsupportedRightShiftOperator = createDiagnosticFactory(
+export const unsupportedRightShiftOperator = createErrorDiagnosticFactory(
     "Right shift operator is not supported for target Lua 5.3. Use `>>>` instead."
 );
 
 const getLuaTargetName = (version: LuaTarget) => (version === LuaTarget.LuaJIT ? "LuaJIT" : `Lua ${version}`);
-export const unsupportedForTarget = createDiagnosticFactory(
+export const unsupportedForTarget = createErrorDiagnosticFactory(
     (functionality: string, version: Exclude<LuaTarget, LuaTarget.Universal>) =>
         `${functionality} is/are not supported for target ${getLuaTargetName(version)}.`
 );
 
-export const unsupportedProperty = createDiagnosticFactory(
+export const unsupportedProperty = createErrorDiagnosticFactory(
     (parentName: string, property: string) => `${parentName}.${property} is unsupported.`
 );
 
-export const invalidAmbientIdentifierName = createDiagnosticFactory(
+export const invalidAmbientIdentifierName = createErrorDiagnosticFactory(
     (text: string) => `Invalid ambient identifier name '${text}'. Ambient identifiers must be valid lua identifiers.`
 );
 
-export const unresolvableRequirePath = createDiagnosticFactory(
+export const unresolvableRequirePath = createErrorDiagnosticFactory(
     (path: string) => `Cannot create require path. Module '${path}' does not exist within --rootDir.`
 );
 
-export const unsupportedVarDeclaration = createDiagnosticFactory(
+export const unsupportedVarDeclaration = createErrorDiagnosticFactory(
     "`var` declarations are not supported. Use `let` or `const` instead."
 );
+
+export const annotationDeprecated = createWarningDiagnosticFactory(
+    (kind: AnnotationKind) =>
+        `'@${kind}' is deprecated and will be removed in a future update. Please update your code before upgrading to the next release, otherwise your project will no longer compile.`
```

Comment:
It might be nice to be able to include some kind of url to documentation on how to upgrade, though it makes this a bit more complicated. Thoughts?

Reply:
Good call. We could link to our annotation docs and add a deprecated label to the annotations in questions. In addition we could add some examples of "how to upgrade" e.g. interface merging for pure abstract...